pycloudlib/lxd/instance.py

```diff
@@ -109,28 +110,36 @@ def ip(self):
         while retries != 0:
             command = [
                 "lxc",
-                "list",
-                "^{}$".format(self.name),
-                "-c4",
-                "--format",
-                "csv",
+                "query",
+                f"/1.0/instances/{self.name}?recursion=1",
             ]
             result = subp(command)
             if result.ok and result.stdout:
-                ip_address = None
                 try:
-                    # Expect "<ip> (<interface>)" when network fully configured
-                    ip_address, _dev = result.stdout.split()
+                    info = json.loads(result.stdout)
                 except ValueError:
                     self._log.debug(
-                        "Unable to parse output of cmd: %s. Expected"
-                        " <ip> (<interface>), got: %s. Retrying %d time(s)...",
+                        "Unable to parse output of cmd: %s. Expected JSON,"
+                        " got: %s. Retrying %d time(s)...",
                         command,
                         result.stdout,
                         retries,
                     )
-                if ip_address:
-                    return ip_address
+                    continue
```

Comment:
@blackboxsw - In case you have a question when reviewing, the continue addition here causes an infinite loop (`retries` is never decremented), so I'm going to change the logic to do what I think this code was supposed to do.

Reply:
The fix (https://github.com/canonical/pycloudlib/pull/221/commits/2cef127657fe5cf9fb7003ff2492f825a837f33a) also refactors parsing logic away from retry logic into `parse_ip()` - this was to make it easier to test parsing logic in isolation easier (and also to make the code a little more readable). Note the `s/continue/else/` in this changeset to fix the exception case. I should have made that into two separate commits to make the review easier.